common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -1509,12 +1515,62 @@ public boolean semanticEquals(final UTF8String other, int collationId) {
     return CollationFactory.fetchCollation(collationId).equalsFunction.apply(this, other);
   }
 
+  private interface SubstringEquals {
+    boolean equals(UTF8String left, UTF8String right, int posLeft, int posRight,
+      int lenLeft, int lenRight);
+  }
+
+  private static class ByteSubstringEquals implements SubstringEquals {
+    @Override
+    public boolean equals(UTF8String left, UTF8String right, int posLeft, int posRight,
+      int lenLeft, int lenRight) {
+      if (lenLeft != lenRight || left.getByte(posLeft) != right.getByte(posRight)) {
+        return false;
+      }
+      return (ByteArrayMethods.arrayEquals(left.base, left.offset + posLeft, right.base,
+        right.offset + posRight, lenLeft));
+    }
+  }
+
+  private static final ByteSubstringEquals byteSubstringEquals = new ByteSubstringEquals();
+
+  private static class CollationSubstringEquals implements SubstringEquals {
+    private final int collationId;
+    private final UTF8String left, right;
+
+    CollationSubstringEquals(int collationId) {
+      this.collationId = collationId;
+      this.left = new UTF8String();
+      this.right = new UTF8String();
+    }
+
+    @Override
+    public boolean equals(UTF8String left, UTF8String right, int posLeft, int posRight,
+      int lenLeft, int lenRight) {
+      moveAddress(this.left, left.base, left.offset + posLeft, lenLeft);
```

Comment:
Should `moveAddress` be an instance method that takes two int parameters (posLeft and lenLeft)?

Reply:
Done, in addition with baseString which is also needed.